### ISSUE TITLE
Can't renew a session !

### ISSUE THREAD

**Issue by AlexPestre** [spotify/ios-sdk#298]
Hi, 

After login to spotify with third party swap, I try to renew the session automaticly after 1 hour.

In my didFinishlaunching method, I have : 

    //Spotify
    id sessionData = [[NSUserDefaults standardUserDefaults] objectForKey:@"spotifySession"];
    SPTSession *sessionUserDefault = [NSKeyedUnarchiver unarchiveObjectWithData:sessionData]?: nil;
    SPTAuth *auth = [SPTAuth defaultInstance];
    auth.tokenRefreshURL = [NSURL URLWithString:kTokenRefreshServiceURL];
    auth.session = sessionUserDefault;
    if (sessionUserDefault) {
        if (![sessionUserDefault isValid] && [auth hasTokenRefreshService]) {
            [auth setSessionUserDefaultsKey:@"spotifySession"];
            [auth renewSession:sessionUserDefault callback:^(NSError *error, SPTSession *session) {
                if (error != nil) {
                    [[NSNotificationCenter defaultCenter] postNotificationName:@"spotifySessionNotOK" object:session];
                }
            }];
        }
        else {
            self.spotifySession = sessionUserDefault;
        }
    }
 
the kTokenRefreshServiceURL is the same url where I replace swap by refresh. 

I receive a 400 error,"No refresh token available in the session!". 

What must I do to renew automaticly the session, please ? 

Alexandre.

**Comment by possan** [comment]
The only difference in our example code is that we set `auth.session = session;` (and `self.spotifySession = session`) in your case, inside the renewSession callback block.

Also, have you set up auth.tokenSwapURL somewhere else?

**Comment by AlexPestre** [comment]
Yep, when I login. 

I tried to use TokenSwapURL or SPTAuthViewController and it’s the same. 

Do I have to get an encryptedRefreshToken after 1st Login ? 

Alexandre.

**Comment by possan** [comment]
Yes, you should get a refresh token after first login.

**Comment by AlexPestre** [comment]
So if not, the first acces was not good, so I have a token for 1 hours only ? 

Alex.

**Comment by possan** [comment]
Yes, the different flows are documented here: https://developer.spotify.com/web-api/authorization-guide/

**Comment by piotrbernad** [comment]
I have the same problem. After first login my encryptedRefreshToken is nil. So it is my fault or sdk?

**Comment by possan** [comment]
Is this still an issue, have you tried the example ruby server included in the latest few betas?

**Comment by AlexPestre** [comment]
no it’s good now
The parse code doesn’t work so I do another solution. 
thanks

**Comment by dnosk** [comment]
my refresh token is nil as well after first login

**Comment by dcristolovean** [comment]
nil refresh token here as well.... beta 10.
this basically makes everything unusable in my app, I can't have the users log in Safari after each hour.  AND having to press OK in that Safari dialog.
No offence, I hope it gets better, but this is by far the worst SDK I ever worked on coming from a "big name" company.

**Comment by yusufkamilak** [comment]
isn't that problem still solved? I am getting nil in refreshToken in first login as well.

**Comment by ketanparekh** [comment]
I am Still getting Error in Swift 3 "No refresh token available in the session!"

**Comment by skylerloth** [comment]
Also having this issue... if anyone has a workaround please post here and help us all

**Comment by psandovalsegura** [comment]
Despite `SPTAuth.defaultInstance().hasTokenRefreshService` and `SPTAuth.defaultInstance().hasTokenSwapService` returning true, my `SPTSession` has a nil `encryptedRefreshToken` so I can't call `renewSession`!

It looks like everyone is having the same issue. Why hasn't this been taken care of?
@possan  @erdtman @JMSerruya
